File: hello-world-examples/go/hello.go
// Hello World in Go - Idiomatic Implementation
// Demonstrates Go best practices and error handling

package main

import (
    "errors"
    "fmt"
    "io"
    "os"
)

// Greeter interface defines greeting behavior
type Greeter interface {
    Greet(name string) (string, error)
}

// SimpleGreeter implements the Greeter interface
type SimpleGreeter struct {
    greeting string
}

// NewGreeter creates a new SimpleGreeter with validation
func NewGreeter(greeting string) (*SimpleGreeter, error) {
    if greeting == "" {
        return nil, errors.New("greeting cannot be empty")
    }
    return &SimpleGreeter{greeting: greeting}, nil
}

// Greet implements the Greeter interface
func (g *SimpleGreeter) Greet(name string) (string, error) {
    if name == "" {
        return "", errors.New("name cannot be empty")
    }
    return fmt.Sprintf("%s, %s!", g.greeting, name), nil
}

// greetToWriter writes greeting to any io.Writer (testable)
func greetToWriter(w io.Writer, greeter Greeter, name string) error {
    message, err := greeter.Greet(name)
    if err != nil {
        return fmt.Errorf("failed to create greeting: %w", err)
    }
    
    _, err = fmt.Fprintln(w, message)
    if err != nil {
        return fmt.Errorf("failed to write greeting: %w", err)
    }
    
    return nil
}

func main() {
    // Simple version
    fmt.Println("Hello, World!")
    
    // Using the struct with error handling
    greeter, err := NewGreeter("Hello")
    if err != nil {
        fmt.Fprintf(os.Stderr, "Error creating greeter: %v\n", err)
        os.Exit(1)
    }
    
    // Greet with error handling
    if err := greetToWriter(os.Stdout, greeter, "World"); err != nil {
        fmt.Fprintf(os.Stderr, "Error: %v\n", err)
        os.Exit(1)
    }
    
    // Custom greeting
    customGreeter, _ := NewGreeter("Greetings")
    if err := greetToWriter(os.Stdout, customGreeter, "Go Developer"); err != nil {
        fmt.Fprintf(os.Stderr, "Error: %v\n", err)
        os.Exit(1)
    }
    
    // Demonstrate error case
    if _, err := greeter.Greet(""); err != nil {
        fmt.Printf("Expected error: %v\n", err)
    }
}

File: hello-world-examples/rust/hello.rs
/// Hello World in Rust - Idiomatic Implementation
/// Demonstrates Rust best practices, error handling, and ownership

use std::error::Error;
use std::fmt;
use std::io::{self, Write};

/// Custom error type for greeting operations
#[derive(Debug)]
enum GreetError {
    EmptyName,
    EmptyGreeting,
    IoError(io::Error),
}

impl fmt::Display for GreetError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            GreetError::EmptyName => write!(f, "Name cannot be empty"),
            GreetError::EmptyGreeting => write!(f, "Greeting cannot be empty"),
            GreetError::IoError(e) => write!(f, "IO error: {}", e),
        }
    }
}

impl Error for GreetError {}

impl From<io::Error> for GreetError {
    fn from(error: io::Error) -> Self {
        GreetError::IoError(error)
    }
}

/// A struct to hold greeting configuration
#[derive(Debug, Clone)]
struct Greeter {
    greeting: String,
}

impl Greeter {
    /// Creates a new Greeter with validation
    fn new(greeting: impl Into<String>) -> Result<Self, GreetError> {
        let greeting = greeting.into();
        if greeting.is_empty() {
            return Err(GreetError::EmptyGreeting);
        }
        Ok(Self { greeting })
    }
    
    /// Creates a greeting message
    fn greet(&self, name: &str) -> Result<String, GreetError> {
        if name.is_empty() {
            return Err(GreetError::EmptyName);
        }
        Ok(format!("{}, {}!", self.greeting, name))
    }
    
    /// Writes greeting to a writer (for testability)
    fn greet_to_writer<W: Write>(
        &self,
        writer: &mut W,
        name: &str,
    ) -> Result<(), GreetError> {
        let message = self.greet(name)?;
        writeln!(writer, "{}", message)?;
        Ok(())
    }
}

impl Default for Greeter {
    fn default() -> Self {
        Self {
            greeting: "Hello".to_string(),
        }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    // Simple version
    println!("Hello, World!");
    
    // Using the struct
    let greeter = Greeter::default();
    greeter.greet_to_writer(&mut io::stdout(), "World")?;
    
    // Custom greeting with error handling
    let custom_greeter = Greeter::new("Greetings")?;
    custom_greeter.greet_to_writer(&mut io::stdout(), "Rust Developer")?;
    
    // Using iterator and functional style
    let names = vec!["Alice", "Bob", "Charlie"];
    names
        .iter()
        .map(|name| greeter.greet(name))
        .collect::<Result<Vec<_>, _>>()?
        .iter()
        .for_each(|msg| println!("{}", msg));
    
    // Demonstrate error handling
    match greeter.greet("") {
        Ok(_) => println!("This shouldn't happen"),
        Err(e) => println!("Expected error: {}", e),
    }
    
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_greet_success() {
        let greeter = Greeter::default();
        assert_eq!(greeter.greet("Test").unwrap(), "Hello, Test!");
    }
    
    #[test]
    fn test_greet_empty_name() {
        let greeter = Greeter::default();
        assert!(matches!(greeter.greet(""), Err(GreetError::EmptyName)));
    }
}